components/builder-core/src/access_token.rs

```diff
@@ -0,0 +1,93 @@
+// Copyright (c) 2017 Chef Software Inc. and/or applicable contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::path::PathBuf;
+use chrono::{self, Utc, Duration};
+
+use hab_net::privilege::FeatureFlags;
+use error::{Error, Result};
+use integrations::{encrypt, decrypt};
+use protocol::{message, sessionsrv};
+
+pub const BUILDER_ACCOUNT_ID: u64 = 0;
+const ACCESS_TOKEN_PREFIX: &'static str = "TOKN";
+const BUILDER_TOKEN_LIFETIME_HOURS: i64 = 2;
+
+pub fn generate_bldr_token(key_dir: &PathBuf) -> Result<String> {
+    generate_access_token(
+        key_dir,
+        BUILDER_ACCOUNT_ID,
+        FeatureFlags::all().bits(),
+        Duration::hours(BUILDER_TOKEN_LIFETIME_HOURS),
+    )
+}
+
+pub fn generate_user_token(key_dir: &PathBuf, account_id: u64, privileges: u32) -> Result<String> {
+    generate_access_token(
+        key_dir,
+        account_id,
+        privileges,
+        Duration::max_value(), // User tokens never expire, can only be revoked
+    )
+}
+
+pub fn generate_access_token(
+    key_dir: &PathBuf,
+    account_id: u64,
+    flags: u32,
+    lifetime: Duration,
+) -> Result<String> {
+    let expires = Utc::now()
+        .checked_add_signed(lifetime)
+        .unwrap_or(chrono::MAX_DATE.and_hms(0, 0, 0))
+        .timestamp();
+
+    let mut token = sessionsrv::AccessToken::new();
+    token.set_account_id(account_id);
+    token.set_flags(flags);
+    token.set_expires(expires);
+
+    let bytes = message::encode(&token).map_err(Error::Protocol)?;
+    let ciphertext = encrypt(key_dir, &bytes)?;
+
+    Ok(format!("{}{}", ACCESS_TOKEN_PREFIX, ciphertext))
+}
+
+pub fn is_access_token(token: &str) -> bool {
+    token.starts_with(ACCESS_TOKEN_PREFIX)
+}
+
+pub fn validate_access_token(key_dir: &PathBuf, token: &str) -> Result<sessionsrv::Session> {
+    assert!(is_access_token(token));
+
+    let encoded = token.to_owned().split_off(ACCESS_TOKEN_PREFIX.len());
+
+    let bytes = decrypt(key_dir, &encoded)?;
```

Comment:
For my rust edification, could this work equally well and obviate the creation of a `String`?
```rust
    let (_, encoded) = token.split_at(ACCESS_TOKEN_PREFIX.len());
    let bytes = decrypt(key_dir, encoded)?;
```

Also, do we want to panic if `validate_access_token` is called with a `token` shorter than `ACCESS_TOKEN_PREFIX`? Presumably that won't happen since the one consumer checks `is_access_token` first, but perhaps the panic should be called out in a doc comment.

Reply:
Oh yes, that's a good catch.. it may still require a to_owned but if not, it's should be perfect - don't know how I missed that... so many string functions.  Re: the panic, yes, that should never happen since we're explicitly checking prefix size earlier..